src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py

```diff
@@ -176,61 +207,214 @@ def __blob_is_not_complete(metadata):
+    if source_location.lower().startswith("git://"):
+        return source_location
+
+
+    response = requests.head(source_location)
+    if response.status_code < 400:
+        return source_location
+    else:
+        raise CLIError("'{}' doesn't exist.".format(source_location))
+
+
+def _check_image_name(image_name):
+    
+    # referenc: https://github.com/docker/distribution/tree/master/reference
+
+    if image_name == "":
+        raise CLIError("'--tag' value should not be empty.")
+    
+    tokens = image_name.split(':')
+    if(len(tokens) > 2):
+        raise CLIError("'--tag' value should be repository-name and optionally a tag in the 'repository-name:tag' format")
+
+    # check repository-name    
+    repository_name = tokens[0]
+    if len(repository_name) > 255:
+        raise CLIError("The repository-name of '--tag' value should be no more than 255 characters.")
     else:
-        # Maybe just raise error?
-        return result
+        if re.match(r"^[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?$", repository_name) is None:
+            raise CLIError("The '--tag' value is not valid. Please check https://docs.docker.com/engine/reference/commandline/tag/.")
+
+    # check tag
+    if len(tokens) == 2:
+        tag = tokens[1]
+        if re.match(r"^[\w][\w.-]{0,127}$", tag) is None:
+            raise CLIError("The '--tag' value is not valid. Please check https://docs.docker.com/engine/reference/commandline/tag/.")
+            
+    return image_name
+
+def _upload_source_code(client, registry_name, resource_group_name, source_location):
+
+    tar_file_path = os.path.join(tempfile.gettempdir(),
+                                 "source_archive_{}.tar.gz".format(hash(os.times())))
+
+    try:
+        logger.debug("Starting to archive the source code to '{}'.".format(tar_file_path))
+
+        # TODO: Check if it works in windows
+        if not source_location.endswith('/'):
+            source_location += '/'
+
+        ignore_list = _load_dockerignore_file(source_location)
+
+        def _filter_file(tarinfo):
+
+            if ignore_list is None:
+                return tarinfo
+                                    
+            for item in ignore_list:
+                if re.match(item.pattern, tarinfo.name): 
+                    logger.debug(".dockerignore: rule '{}' matches '{}'.".format(item.rule, tarinfo.name))
+                    return None if item.ignore else tarinfo
+
+            logger.debug(".dockerignore: no rule for '{}'.".format(tarinfo.name))
+            return tarinfo
+
+        with tarfile.open(tar_file_path, "w:gz") as tar:
+            tar.add(source_location, arcname=os.path.basename(source_location), filter=_filter_file)
+
+        logger.debug("Starting to upload the archived source code from '{}'.".format(tar_file_path))
+
+        source_upload_location = client.get_source_upload_url(
```

Comment:
we should call this one before making the `tar`, it would be really sad to have a transient error here after spending a few mins making the tar - the block blob service should have retries automatically i believe

Reply:
updated.